### PR TITLE
Multiple Gather to Split Fusion

### PR DESCRIPTION
**Pull request by iK1D**
For below code in some transformers models:
```
fused_qkv = fused_qkv.view(batch_size, seq_length, self.num_heads, 3, self.head_dim)
return fused_qkv[..., 0, :], fused_qkv[..., 1, :], fused_qkv[..., 2, :]
```

The exported graph will contains 3 Gather nodes, currently ORT's GatherGrad CUDA implementation is slow. This pattern can be fused to use one Split, so that we can launch less kernels for the compute, the perf of Split/Concat (for grad) is also better than Gather/GatherGrad.

In a real example, one GatherGrad will take 15ms and there are 3 for each layer in the graph, after the fusion, one Concat takes only 35us. The total time of a step is improved from 1.5s to 0.4s.

### CODE REVIEW

File: orttraining/orttraining/core/optimizer/graph_transformer_utils.cc [hunk 1]
```diff
@@ -99,6 +100,7 @@ std::vector<std::unique_ptr<GraphTransformer>> GeneratePreTrainingTransformers(
       transformers.emplace_back(std::make_unique<SimplifiedLayerNormFusion>(compatible_eps));
       transformers.emplace_back(std::make_unique<FastGeluFusion>(compatible_eps));
       transformers.emplace_back(std::make_unique<SoftmaxCrossEntropyLossInternalFusion>(compatible_eps));
+      transformers.emplace_back(std::make_unique<GatherToSplitFusion>(compatible_eps));
```
Comment (askhade) [inline orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:103]: When are the transformers under "GeneratePreTrainingTransformers" applied?

Comment (askhade) [reply orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:103]: looks like they are applied as level 1 transformers. am I right?

**Review by prathikr — APPROVED** [review]
Baseline Test Command: python -m torch.distributed.launch --nproc_per_node=8 workspace/transformers/examples/pytorch/language-modeling/run_clm.py --model_name_or_path bigscience/bloom-560m --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_train --output_dir /dev/shm --overwrite_output_dir --max_steps 200 --logging_steps 20 --per_device_train_batch_size 2 --fp16

Optimum Test Command: python -m torch.distributed.launch --nproc_per_node=8 workspace/optimum/examples/onnxruntime/training/language-modeling/run_clm.py --model_name_or_path bigscience/bloom-560m --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_train --output_dir /dev/shm --overwrite_output_dir --max_steps 200 --logging_steps 20 --per_device_train_batch_size 2 --fp16

Total Training Time Results
Baseline: 1.5min
Optimum: 6min
Optimum w/ gather-to-split fusion: 2min

**Review by askhade — APPROVED** [review]
LGTM

**Comment by askhade** [comment]
> Baseline Test Command: python -m torch.distributed.launch --nproc_per_node=8 workspace/transformers/examples/pytorch/language-modeling/run_clm.py --model_name_or_path bigscience/bloom-560m --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_train --output_dir /dev/shm --overwrite_output_dir --max_steps 200 --logging_steps 20 --per_device_train_batch_size 2 --fp16
> 
> Optimum Test Command: python -m torch.distributed.launch --nproc_per_node=8 workspace/optimum/examples/onnxruntime/training/language-modeling/run_clm.py --model_name_or_path bigscience/bloom-560m --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_train --output_dir /dev/shm --overwrite_output_dir --max_steps 200 --logging_steps 20 --per_device_train_batch_size 2 --fp16
> 
> Total Training Time Results Baseline: 1.5min Optimum: 6min Optimum w/ gather-to-split fusion: 2min

Interesting so we are still slower than stock pytorch for bloom?

File: orttraining/orttraining/core/optimizer/graph_transformer_utils.cc (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/core/optimizer/graph_transformer_utils.cc:103]: They are applied before gradient builder for training case. Normally if you want it take effect before gradient builder so that gradient builder can use it you need to add it here. E.g., we fuse Gather nodes to Split, then gradient builder will generate Split's gradient, instead of Gather's gradients. You may also need to add it to the file in onnxruntime folder so that it can also help the inference scenario.

Yes, it's at level 1.

**Comment by iK1D** [comment]
> > Baseline Test Command: python -m torch.distributed.launch --nproc_per_node=8 workspace/transformers/examples/pytorch/language-modeling/run_clm.py --model_name_or_path bigscience/bloom-560m --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_train --output_dir /dev/shm --overwrite_output_dir --max_steps 200 --logging_steps 20 --per_device_train_batch_size 2 --fp16
> > Optimum Test Command: python -m torch.distributed.launch --nproc_per_node=8 workspace/optimum/examples/onnxruntime/training/language-modeling/run_clm.py --model_name_or_path bigscience/bloom-560m --dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_train --output_dir /dev/shm --overwrite_output_dir --max_steps 200 --logging_steps 20 --per_device_train_batch_size 2 --fp16
> > Total Training Time Results Baseline: 1.5min Optimum: 6min Optimum w/ gather-to-split fusion: 2min
> 
> Interesting so we are still slower than stock pytorch for bloom?

If total time means E2E time, it's not fair for ORT as ORT will take long time for preparation in the 1st step, especially just running 1 or 2 mins. It's better to measure the perf when the running is relatively stable.

**Comment by askhade** [comment]
@prathikr  can you exclude the first step and compare the throughput? Also can you try the max batch size exercise and see how much can we push?

**Comment by prathikr** [comment]
<html xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:dt="uuid:C2F41010-65B3-11d1-A29F-00AA00C14882"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=OneNote.File>
<meta name=Generator content="Microsoft OneNote 15">
</head>

<body lang=en-US style='font-family:Calibri;font-size:11.0pt'>


<p style='margin:0in;font-family:Calibri;font-size:11.0pt'>Baseline Command:
python workspace/transformers/examples/pytorch/language-modeling/run_clm.py
--model_name_or_path bigscience/bloom-560m --dataset_name wikitext
--dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_train
--output_dir /dev/shm --overwrite_output_dir --max_steps 500 --logging_steps 50
--per_device_train_batch_size 2 --fp16</p>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt'>&nbsp;</p>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt'>Optimum Command:
python
workspace/optimum/examples/onnxruntime/training/language-modeling/run_clm.py
--model_name_or_path bigscience/bloom-560m --dataset_name wikitext
--dataset_config_name wikitext-2-raw-v1 --label_smoothing 0.1 --do_train
--output_dir /dev/shm --overwrite_output_dir --max_steps 500--logging_steps 50
--per_device_train_batch_size 2 --fp16</p>

<p style='margin:0in;font-family:Calibri;font-size:11.0pt'>&nbsp;</p>

<div style='direction:ltr'>



  | Epoch | Train_loss | Train_runtime   (H:MM:SS) | Train_samples | Train_samples_per_second | Train_steps_per_second | Per_device_train_batch_size
-- | -- | -- | -- | -- | -- | -- | --
Baseline | 0.43 | 4.7208 | 0:03:41 | 2318 | 4.524 | 2.262 | 2
Optimum | 0.43 | 4.7202 | 0:13:04 | 2318 | 1.274 | 0.637 | 2
Optimum w/ multiple_gather_to_split | 0.43 | 4.7173 | 0:03:48 | 2318 | 4.368 | 2.184 | 2
Baseline | OOM :( |   |   |   |   |   | 4
Optimum w/ multiple_gather_to_split | OOM :( |   |   |   |   |   | 4



</div>


</body>

</html>
